docs/content/specs/terraform/_index.md

```diff
@@ -533,6 +543,466 @@ Bad examples:
+}
+```
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR27 - Category: Code Style - Declaration of a provider in the module
+
+[By rules](https://www.terraform.io/docs/language/modules/develop/providers.html), in the module code `provider` cannot be declared. The only exception is when the module indeed need different instances of the same kind of `provider`(Eg. manipulating resources across different `location`s or accounts), you **MUST** declare `configuration_aliases` in `terraform.required_providers`. See details in this [document](https://www.terraform.io/docs/language/providers/configuration.html#alias-multiple-provider-configurations).
+
+`provider` block declared in the module can only be used to differentiate instances used in  `resource` and `data`. Declaration of fields other than `alias` in `provider` block is strictly forbidden. It could lead to module users unable to utilize `count`, `for_each` or `depends_on`. Configurations of the `provider` instance should be passed in by the module users.
+
+Good examples:
+
+In verified module:
+
+```terraform
+terraform {
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "~> 3.0"
+      configuration_aliases = [ azurerm.alternate ]
+    }
+  }
+}
+```
+
+In the root module where we call this verified module:
+
+````terraform
+provider "azurerm" {
+  features {}
+}
+
+provider "azurerm" {
+  alias = "alternate"
+  features {}
+}
+
+module "foo" {
+  source = "xxx"
+  providers = {
+    azurerm = azurerm
+    azurerm.alternate = azurerm.alternate
+  }
+}
+````
+
+Bad example:
+
+In verified module:
+
+```terraform
+provider "azurerm" {
+  # Configuration options
+  features {}
+}
+```
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR28 - Category: Code Style - `output` **MUST** be arranged alphabetically
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR29 - Category: Code Style - `output` contains confidential data should declare `sensitive = true`
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR30 - Category: Code Style - Dealing with Deprecated `output`s
+
+Sometimes we notice that the name of certain `output` is not appropriate anymore, however, since we have to ensure forward compatibility in the same major version, it's not allowed to change the name directly. We need to move it to an independent `deprecated-outputs.tf` file, then redefine a new output in `output.tf` and make sure it's compatible everywhere else in the module.
+
+A cleanup can be performed to `deprecated-outputs.tf` and other logics related to compatibility during a major version upgrade.
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR31 - Category: Code Style - `locals.tf` **MUST** contain only one `locals` block
+
+All `local` **MUST** be defined in the only `locals` block in `locals.tf` file.
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR32 - Category: Code Style - `local` should be arranged alphabetically
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR33 - Category: Code Style - `local` should use types as precise as possible
```

Comment:
We cannot use strict types for locals. I think this needs expanding as the `map(string)` will be confusing for people.

I think this needs context for the use of the local variable, typically from a resource or variable input.